### PR TITLE
network: implement transaction request handling.

### PR DESCRIPTION
**Pull request by hdevalence**
This commit makes several related changes to the network code:

- adds a `TransactionsByHash(HashSet<transaction::Hash>)` request and
  `Transactions(Vec<Arc<Transaction>>)` response pair that allows
  fetching transactions from a remote peer;

- adds a `PushTransaction(Arc<Transaction>)` request that pushes an
  unsolicited transaction to a remote peer;

- adds an `AdvertiseTransactions(HashSet<transaction::Hash>)` request
  that advertises transactions by hash to a remote peer;

- adds an `AdvertiseBlock(block::Hash)` request that advertises a block
  by hash to a remote peer;

Then, it modifies the connection state machine so that outbound
requests to remote peers are handled properly:

- `TransactionsByHash` generates a `getdata` message and collects the
  results, like the existing `BlocksByHash` request.

- `PushTransaction` generates a `tx` message, and returns `Nil` immediately.

- `AdvertiseTransactions` and `AdvertiseBlock` generate an `inv`
  message, and return `Nil` immediately.

Next, it modifies the connection state machine so that messages
from remote peers generate requests to the inbound service:

- `getdata` messages generate `BlocksByHash` or `TransactionsByHash`
  requests, depending on the content of the message;

- `tx` messages generate `PushTransaction` requests;

- `inv` messages generate `AdvertiseBlock` or `AdvertiseTransactions`
  requests.

Finally, it refactors the request routing logic for the peer set to
handle advertisement messages, providing three routing methods:

- `route_p2c`, which uses p2c as normal (default);
- `route_inv`, which uses the inventory registry and falls back to p2c
  (used for `BlocksByHash` or `TransactionsByHash`);
- `route_all`, which broadcasts a request to all ready peers (used for
  `AdvertiseBlock` and `AdvertiseTransactions`).

Extracted from #1008.

### CODE REVIEW

**Comment by hdevalence** [comment]
Ooops, I split these too finely, blocked on #1013.

File: zebra-network/src/peer/connection.rs
Comment (yaahc) [inline zebra-network/src/peer/connection.rs:451]: whats the difference between `tx` and `peer_tx` here?

File: zebra-network/src/peer/connection.rs
Comment (yaahc) [inline zebra-network/src/peer/connection.rs:487]: do we want to try to find a way to get exhaustive matching here?

File: zebra-network/src/peer/connection.rs
Comment (yaahc) [inline zebra-network/src/peer/connection.rs:601]: you could collapse this by using `collect` and some error handling internally, this changes the behaviour a bit, so it no longer prints "unhandled message type" if any of them would map to none, you could probably add a `.map_err` before the `ok` to print a similar message if it's important.

```suggestion
            Message::Inv(items) => items
                .iter()
                .map(|item| {
                    if let InventoryHash::Tx(hash) = item {
                        Ok(*hash)
                    } else {
                        Err(())
                    }
                })
                .collect::<Result<_, ()>>()
                .map(Request::AdvertiseTransactions)
                .ok(),
```

File: zebra-network/src/peer_set/set.rs
Comment (yaahc) [inline zebra-network/src/peer_set/set.rs]: I'm gonna turn this into a clippy issue, false positives are no bueno, however it seems like `take` is more appropriate here.
```suggestion
        let ready_services = std::mem::take(&mut self.ready_services);
```

File: zebra-network/src/peer_set/set.rs
Comment (yaahc) [inline zebra-network/src/peer_set/set.rs]: i think the `into_iter` is implied by the for loop

```suggestion
        for (key, mut svc) in ready_services {
```

File: zebra-network/src/peer/connection.rs
Comment (hdevalence) [reply zebra-network/src/peer/connection.rs:487]: Yeah, the way we'd do that is by combining all the match statements into one, but I didn't want to do it in this change because it's already quite large.

File: zebra-network/src/peer_set/set.rs
Comment (hdevalence) [reply zebra-network/src/peer_set/set.rs]: Oh, take is much better!

File: zebra-network/src/peer/connection.rs
Comment (yaahc) [reply zebra-network/src/peer/connection.rs:487]: okay, wanna make an issue for this assuming there isn't one already?

File: zebra-network/src/protocol/internal/request.rs
Comment (teor2345) [inline zebra-network/src/protocol/internal/request.rs]: Isn't the size a requirement?
```suggestion
    /// block. This routing is only used for request sets of size 1.
```

File: zebra-network/src/protocol/internal/request.rs
Comment (teor2345) [inline zebra-network/src/protocol/internal/request.rs]: Similarly, isn't the size a requirement?
```suggestion
    /// the transaction. This routing is only used for request sets of size 1.
```

**Comment by teor2345** [comment]
>  * `inv` messages generate `AdvertiseBlock` or `AdvertiseTransactions`
>     requests.

Does the inbound service handle `AdvertiseBlock` by fetching and verifying the block?
If not, is there a ticket or checklist item for this feature?

**Review by teor2345 — CHANGES_REQUESTED** [review]
Overall, this looks pretty good.

What's the test plan here?

File: zebra-network/src/peer/connection.rs
Comment (hdevalence) [reply zebra-network/src/peer/connection.rs:451]: https://github.com/ZcashFoundation/zebra/pull/1016/files/409a2289b57aa71cacbf19b17f6f5839890092e3#diff-238888844b1b8121c9450ba2070242beR375

`tx` is the oneshot used to send the response to the request, `peer_tx` is the outbound message sink.

File: zebra-network/src/peer/connection.rs
Comment (hdevalence) [reply zebra-network/src/peer/connection.rs:487]: Yep, I'll do that after this is merged.

File: zebra-network/src/peer/connection.rs
Comment (hdevalence) [reply zebra-network/src/peer/connection.rs:601]: I think it's important that the match arm matches *only* messages where all of the inventory items are transaction hashes, because otherwise we're moving match logic out of the match statement and into the arms.  The code is superficially uglier, yes, but it's more cleanly factored, because all of the matching criteria are done upfront and the match arm is independent of other match arms.

File: zebra-network/src/protocol/internal/request.rs
Comment (hdevalence) [reply zebra-network/src/protocol/internal/request.rs]: Yes, good catch, I wrote the comment before dropping the more complex logic.

**Comment by hdevalence** [comment]
> > * `inv` messages generate `AdvertiseBlock` or `AdvertiseTransactions`
> >   requests.
> 
> Does the inbound service handle `AdvertiseBlock` by fetching and verifying the block?
> If not, is there a ticket or checklist item for this feature?

Yes, that's part of #889.

> What's the test plan here?

In a later chunk of work, we should build two `peer::Connections` hooked up to each other, send messages through one, and check that they come out as expected on the other side.  Until now we haven't done any real message<->request plumbing, except as was necessary to generate outbound requests, which is why this harness doesn't exist yet.

**Comment by hdevalence** [comment]
Rebased onto `main` to pick up #1013.

File: zebra-network/src/peer/connection.rs
Comment (yaahc) [reply zebra-network/src/peer/connection.rs:601]: makes sense

**Review by yaahc — APPROVED** [review]
Hard to tell exactly what changes have been made since the last review since you force pushed and its all one commit so I'm going to go ahead and just approve.

**Comment by hdevalence** [comment]
@yaahc There's no changes except the suggestions, which are recorded above.

**Comment by yaahc** [comment]
perfect, lgtm

**Review by teor2345 — APPROVED** [review]
Looks good to me.

Just a reminder about the follow-up work:
* the way we'd do that is by combining all the match statements into one https://github.com/ZcashFoundation/zebra/pull/1016#discussion_r484535408
* In a later chunk of work, we should build two peer::Connections hooked up to each other, send messages through one, and check that they come out as expected on the other side https://github.com/ZcashFoundation/zebra/pull/1016#issuecomment-688460593

**Comment by hdevalence** [comment]
Follow-up work to be done in #1048.

**Pull request merged**
